### PR TITLE
limited max amount of components in circuits

### PR DESCRIPTION
**Pull request by PowerfulBacon**
## About The Pull Request

Limits certain components maximum amount in cicrcuits to prevent certain components which are highly abusable if multiple are put in at once.

## Why It's Good For The Game

Things that can instantly kill someone after 3 seconds, instantly down someone aren't fun except for the person using them. (I must say I had a lot of fun using them though)

## Changelog
:cl:
tweak: Abusable circuit components have been given a limit to how many can be out in a circuit.
/:cl:

### CODE REVIEW

**Comment by St0rmC4st3r** [comment]
Mmmm. I don't really like it. Maybe make it only be able to draw blood from living while they are horizontal?

**Comment by ike709** [comment]
👎 

Make a subtype circuit specifically for drawing blood with a high cooldown and a very low max transfer. And maybe include a do_after so they have to stand still?

**Comment by zeskorion** [comment]
Nah. This is a symptom, not the root cause. What needs to be done is a cap on certain elements (gun holder,  injector, locomotion circuit, etc) to prevent dumb exploits like OHKO lasers, lightspeed drones, and insta exsanguination

**Comment by PowerfulBacon** [comment]
> 
> 
> Nah. This is a symptom, not the root cause. What needs to be done is a cap on certain elements (gun holder, injector, locomotion circuit, etc) to prevent dumb exploits like OHKO lasers, lightspeed drones, and insta exsanguination

I was going to do this, I guess this is the solution I'll take since things like 6 shot disablers that instakill are kinda borked

**Comment by PowerfulBacon** [comment]
@ike709 @zeskorion 
Implemented zesko's solution adding a limit to certain components.
This adds limits to:
 - Lights (Max 6) | Reason: Easy to lag the server with a lot of lights
 - Sound (Max 5 or 6) | Reason: Playing like 20 sounds at the same time makes it earrape
 - Weapons (Max 1) | Reason: Shooting 6 bullets in 1 is overpowered
 - Injector (Max 1) | Instakill injectors are being abused
 - Some other stuff that would be broken if spammed but I can't remember

**Review by zeskorion — DISMISSED** [review]
looks solid to me

File: code/modules/integrated_electronics/core/assemblies.dm [hunk 1]
```diff
@@ -457,6 +457,15 @@
 	var/total_part_size = return_total_size()
 	var/total_complexity = return_total_complexity()
 
+	if(IC.max_allowed)
+		var/current_components
+		for(var/obj/item/integrated_circuit/component in assembly_components)
+			if(component.type == IC.type)
+				current_components ++
```
Comment (ike709) [inline code/modules/integrated_electronics/core/assemblies.dm]:
```suggestion
				current_components++
```

File: code/modules/integrated_electronics/core/assemblies.dm [hunk 2]
```diff
@@ -457,6 +457,15 @@
 	var/total_part_size = return_total_size()
 	var/total_complexity = return_total_complexity()
 
+	if(IC.max_allowed)
+		var/current_components
+		for(var/obj/item/integrated_circuit/component in assembly_components)
```
Comment (ike709) [inline code/modules/integrated_electronics/core/assemblies.dm]: Can this be typeless?

Comment (PowerfulBacon) [reply code/modules/integrated_electronics/core/assemblies.dm]: Yea sure

Comment (PowerfulBacon) [inline code/modules/integrated_electronics/core/assemblies.dm]:
```suggestion
		for(var/obj/item/integrated_circuit/component as anything in assembly_components)
```

**Comment by RafRoq** [comment]
never died from this neither saw someone dying from this, i will remain neutral, but i dont like the idea of just removing the feature, what about adding a delay ?

**Comment by zeskorion** [comment]
> never died from this neither saw someone dying from this, i will remain neutral, but i dont like the idea of just removing the feature, what about adding a delay ?

It already has one, but exsanguination is a very difficult condition to recover from

**Comment by ghost** [comment]
Right now it takes time for the circuit to syringe - you have to let it start and not move while it's happening. I think you should just add a loading bar above the circuit instead of removing features.

**Comment by PowerfulBacon** [comment]
PR name was outdated, this doesn't remove anything, it adds a max amount of certain components per circuit meaning you cant have a circuit with 6 gun firers or like 6 injectors allowing you to almost instakill someone if they stand still

**Comment by PowerfulBacon** [comment]
> never died from this neither saw someone dying from this, i will remain neutral, but i dont like the idea of just removing the feature, what about adding a delay ?

We had someone use it to fuck a wizard earlier, the only message that pops up warns you but the time you read it, it has already taken all of your blood (it takes way too much out, 30 units at a time + you can put multiple in 1 circuit allowing you to remove hundreds of units of blood from someone at a time)

**Comment by sheldoniscool** [comment]
> Injector = 1
This destroys some of my smart chem auto injectos.

The thing is - I can make two syringes load them into a syringe gun and instakill someone in a moment. Like that's the bar. I think there's nothing wrong with it either. There's many defences to the syringe gun.

You'd have to not move in order to have your blood sucked, maybe make it so syringes cant go through things that normal syringes cant?

I've yet to see any issues with lag. I think this is the trade off with circuits, and honestly circuits is a big reason I play bee.
Circuits come with infinite possiblities. By attempting to cull and control you are limiting what you can do with them and once two or one syringe from a syringe gun can be a death sentence so yeah.

It's rare to see them abused. I'd rather admins be like 'hey you've done that a few time's now how about you try something else'

**Comment by sheldoniscool** [comment]
Also to add - I think balancing should only be done when needed - This has survived the test of time. Make cool new things.

**Comment by sheldoniscool** [comment]
> > never died from this neither saw someone dying from this, i will remain neutral, but i dont like the idea of just removing the feature, what about adding a delay ?
> 
> We had someone use it to fuck a wizard earlier, the only message that pops up warns you but the time you read it, it has already taken all of your blood (it takes way too much out, 30 units at a time + you can put multiple in 1 circuit allowing you to remove hundreds of units of blood from someone at a time)

Perhaps have it have text popup above it just like when people speak saying 'beginning to inject/drawblood' or the loading bar idea seem like a proper fix.

This should be voted on by the players as opposed to forced in by two maintainers. I guarantee players won't want it.

**Comment by zeskorion** [comment]
> > Injector = 1
> > This destroys some of my smart chem auto injectos.
> 
> The thing is - I can make two syringes load them into a syringe gun and instakill someone in a moment. Like that's the bar. I think there's nothing wrong with it either. There's many defences to the syringe gun.
> 
> You'd have to not move in order to have your blood sucked, maybe make it so syringes cant go through things that normal syringes cant?
> 
> I've yet to see any issues with lag. I think this is the trade off with circuits, and honestly circuits is a big reason I play bee.
> Circuits come with infinite possiblities. By attempting to cull and control you are limiting what you can do with them and once two or one syringe from a syringe gun can be a death sentence so yeah.
> 
> It's rare to see them abused. I'd rather admins be like 'hey you've done that a few time's now how about you try something else'

there is a pr up to fix the syringe thing. Meme instakill edge cases should be removed wherever possible

**Review by zeskorion — APPROVED** [review]

**Comment by AnjelicaChaykovsky** [comment]
Weapons (Max 1) | Reason: Shooting 6 bullets in 1 is overpowered

i personally never did that, i didnt event fool around with guns that much,  all i did was use gattling gun which goes pew pew pew and then stop,  it shoots 3 times and is so fun to use.
What made you think "max 1" Is  reasonable ? Just remove it at this point.... this assembly cant even recharge

the only thing that should be removing  is the game breaking  stuff why remove fun gattling guns 


i also have 3 part complex device  and the injection update changed it a bit i thought it was always like that but now i think that before you could move while being injected ? 
Also while im talking about my health system its a shame that reagent filter circuit doesnt work kek  currently above 10 brute/burn it injects you with both burn/brute chems but it was supposed to be big brain device not i waste half chems device.
Ofc i could possibly filter reagents without using the reagent filter, but complexity is at 150/150  and its fken one part there are two more.

Also  is there circuit bug report ? I have found so many

**Comment by PowerfulBacon** [comment]
> this assembly cant even recharge

Use advanced energy guns

The main reason for this is that they are kind of broken and can easilly break the server. I have a bot that will go towards the nearest human and suck out all of their blood in 3 seconds which took about 15 minutes to make

**Comment by AnjelicaChaykovsky** [comment]
> Use advanced energy guns

yes but that takes time, also the fact that guns are limited, makes the gun circuit spam far less broken compared to average joe roundstart instakill machine

anyway  i wanted to ask you if u could change beeper circuit  ? It has 8 complexity  which is the worst thing ever, instead it could have a resonable 1 complexity but only couple can fit into assembly, 8 complexity is such a waste, my machines always have complexity maxed out so i dont use beeper circuits... i want to though.

**Review by zeskorion — CHANGES_REQUESTED** [review]
before this is merged, i've been informed of another exploit; please cap extinguisher circuits, as well as capping the circuit reagent heater (the circuit heater is the only uncapped reagent heater i'm aware of, with the next highest method being an advanced welding torch at 3800 max. put it down to around 3000 max?)

**Review by zeskorion — APPROVED** [review]

**Comment by eLeCtrOssSnake** [comment]
why...

**Comment by PowerfulBacon** [comment]
> why...

Circuits are highly exploitable and cause a lot of lag on the server. It is very easy to make overpowered machines that can instakill or create lag machines

**Pull request merged**